dial: add `DialContext` function

DerekBum opened the pull request:
In order to replace timeouts with contexts in `Connect` instance creation (go-tarantool), I need a `DialContext` function. It accepts context, and cancels, if context is canceled by user.

Disabled flaking CI Windows runner. Because we don't support connector on Windows, I considered disabling these tests.

Part of https://github.com/tarantool/go-tarantool/issues/136

DerekBum committed:
dial: add `DialContext` function

In order to replace timeouts with contexts in `Connect` instance
creation (go-tarantool), I need a `DialContext` function.
It accepts context, and cancels, if context is canceled by user.

Part of https://github.com/tarantool/go-tarantool/issues/136

oleg-jukovec commented on net.go:
```diff
@@ -109,11 +122,11 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 
-func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
-	session []byte) (*Conn, error) {
+func dialSession(d net.Dialer, context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags, session []byte) (*Conn, error) {
```
Let's stop adding unnecessary arguments here and reorganize the function a little.

For example, it could be something like this:

1.
```suggestion
func createSession(c net.Conn, ctx *Ctx, flags DialFlags, session []byte) (*Conn, error) {
```

2. A separate function for `ctx *Ctx` preparation:
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L134-L141

3. A separate function for `c net.Conn` creation:
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L130-L133
and
https://github.com/tarantool/go-openssl/blob/bdadb815fbb91fdee2b475f7f38b17a7f56db0e0/net.go#L143-L146

or even
```suggestion
func setSession(client *Conn, flags DialFlags, session []byte) (*Conn, error) {
```

oleg-jukovec replied:
Or you can split this function in some other way.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -90,7 +91,19 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+//
+// See func Dial for a description of the network, addr, ctx and flags
+// parameters.
+func DialContext(context context.Context, network, addr string,
```
It would be nice to see a simple test for the call if it's not too complicated.

DerekBum replied:
New tests added into `net_test.go`.

oleg-jukovec commented on net.go:
```diff
@@ -90,7 +91,33 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
 	flags DialFlags) (*Conn, error) {
 	d := net.Dialer{Timeout: timeout}
-	return dialSession(d, network, addr, ctx, flags, nil)
+	if err := prepareCtx(ctx); err != nil {
```
The value `ctx` will not be updated.
```suggestion
	if ctx, err := prepareCtx(ctx); err != nil {
```

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
+}
 
-	c, err := d.Dial(network, addr)
+func createConnection(d net.Dialer, context context.Context, network,
+	addr string) (net.Conn, error) {
```
`context.Context` should be always first arg.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 	if flags&DisableSNI == 0 {
 		err = conn.SetTlsExtHostName(host)
 		if err != nil {
 			conn.Close()
```
We open the connection out of the scope of the function. So let and close it out of the scope in the error case.

This will make it easier to understand what is happening in the error case from the caller code.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -109,58 +136,81 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
+	}
+	conn, err := Client(c, ctx)
+	if err != nil {
+		c.Close()
```
The same about `c` and the `createSession`.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -90,7 +91,37 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
Nit, I don't sure the close of the network connection at the end of the call is a good idea. It should not work.

DerekBum replied:
Thanks, updated.

oleg-jukovec commented on net.go:
```diff
@@ -90,7 +91,37 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+		return nil, err
+	}
+	defer conn.Close()
+	return createSession(conn, flags, addr, ctx, nil)
```
```suggestion
	c, err := createSession(conn, flags, addr, ctx, nil)
	if err != nil {
	    conn.Close()
	    return nil, err
	}
	return c, nil
```

DerekBum replied:
Thanks, updated.

oleg-jukovec commented on net.go:
```diff
@@ -126,41 +161,58 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
```
It closes the connection at the end of the function. But we need an open connection.

Please, update tests to handle the situation.

DerekBum replied:
New test added.

oleg-jukovec commented on net.go:
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
+	return c, nil
+}
+
+func applyFlags(conn *Conn, host string, flags DialFlags) error {
```
```suggestion
func handshake(conn *Conn, host string, flags DialFlags) error {
```

The essence of the function not about flags applying.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
+
+func createSession(c net.Conn, flags DialFlags, addr string, ctx *Ctx,
+	session []byte) (*Conn, error) {
+	host, _, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
```
The code is about a network connection/possibility of a network connection. Not about the create TLS session from the connection.

It is better to move it into `createConnection()` or a separate call.

```suggestion
func createSession(c net.Conn, flags DialFlags, host string, ctx *Ctx,
	session []byte) (*Conn, error) {
```

DerekBum replied:
Updated.

oleg-jukovec commented on net_test.go:
```diff
@@ -0,0 +1,77 @@
+	if err := ctx.SetCipherList("AES128-SHA"); err != nil {
+		t.Fatal(err)
+	}
+	ssl_listener, err := Listen("tcp", "localhost:0", ctx)
```
See linter errors.

DerekBum replied:
Fixed.

oleg-jukovec commented on net_test.go:
```diff
@@ -0,0 +1,77 @@
+	}
+	ssl_listener, err := Listen("tcp", "localhost:0", ctx)
+
+	go sslConnect(t, ssl_listener)
```
It would be nice to wait for the end of the gorouitine execution (with WaitGroup or a channel).

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -126,41 +170,59 @@ func dialSession(d net.Dialer, network, addr string, ctx *Ctx, flags DialFlags,
+}
 
-	c, err := d.Dial(network, addr)
+func createConnection(context context.Context, d net.Dialer, network,
+	addr string) (net.Conn, error) {
+	c, err := d.DialContext(context, network, addr)
```
It could be simplified and unified in such way:
```suggestion
func createConnection(context context.Context, network, addr string) (net.Conn, error) {
	return net.Dialer{}.DialContext(context, network, addr)
```
and usage in the timeout case (`DialTimeout`) will be:
```
ctx, cancel := context.WithTimeout(context.Background(), timeout)
defer cancel()
conn, err := createConnection(ctx, network, addr)
```

Or you could even don't create a separate function and just use different `net.*` calls for different cases.
`Dial()/DialTimeout()/DialContext()` -> `net.Dial()/net.DialTimeout()/net.Dialer{}.DialContext()`.

DerekBum replied:
Updated. Now I use `net.*` calls and host parsing is in a separate function now.

oleg-jukovec commented on net.go:
```diff
@@ -90,7 +91,43 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+func DialContext(context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags) (*Conn, error) {
+	d := net.Dialer{}
+	ctx, err := prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(context, d, network, addr)
+	if err != nil {
+		return nil, err
+	}
```
It is very likely that a connection will fail, so we will not need a ssl context object.
```suggestion
	conn, err := createConnection(context, d, network, addr)
	if err != nil {
		return nil, err
	}
	ctx, err := prepareCtx(ctx)
	if err != nil {
		return nil, err
	}
```

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -89,8 +90,53 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+	ctx, err = prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err != nil {
		conn.Close()
		return nil, err
	}
```

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -89,8 +90,53 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+	ctx, err = prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
```
```suggestion
	if err != nil {
		conn.Close()
		return nil, err
	}
```

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -108,59 +154,80 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 		return nil, err
 	}
-	conn, err := Client(c, ctx)
+	ctx, err = prepareCtx(ctx)
```
+conn.Close() as the same.

DerekBum replied:
Updated.

oleg-jukovec commented on net.go:
```diff
@@ -108,59 +154,80 @@ func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 		}
+		// TODO: use operating system default certificate chain?
 	}
+	return ctx, nil
```
```suggestion
if ctx == nil {
    return NewCtx()
}
return ctx, nil
```

DerekBum replied:
Updated.

oleg-jukovec commented on net_test.go:
```diff
@@ -0,0 +1,94 @@
+package openssl
```
Please, don't test the private API, just a public API behavior.
```suggestion
package openssl_test
```

DerekBum replied:
Updated.

oleg-jukovec commented on net_test.go:
```diff
@@ -0,0 +1,94 @@
+	if err != nil {
+		t.Fatalf("unexpected err: %v", err)
+	}
+	if client.is_shutdown {
+		t.Fatal("client is closed after creation")
+	}
```
Let's better send a few bytes and read it from the echo server (It should work with sslConnect).

DerekBum replied:
Updated.

oleg-jukovec commented:
Please, disable the windows runners (it is not our target platform or we need to fix tests on it).

oleg-jukovec commented on net.go:
```diff
@@ -89,8 +90,55 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
+//
+// See func Dial for a description of the network, addr, ctx and flags
+// parameters.
+func DialContext(context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags) (*Conn, error) {
```
I would also like to suggest this renaming for the whole file:

1. There is now shadowing of the `context` module with the `context` variable.
2. On `Go` a variable `ctx` is usually `context.Context`. So the code would be clearer for a outside reader.
```suggestion
func DialContext(ctx context.Context, network, addr string,
	sslCtx *Ctx, flags DialFlags) (*Conn, error) {
```

DerekBum replied:
Updated.

oleg-jukovec reviewed (approved):
Thank you for the patch!

DerekBum committed:
CI: disable Windows runner

Disabling of the Windows runner. These tests were flaking.
Because we don't support connector on Windows, I considered
disabling these tests.

Part of #10

DerekBum commented:
Disabled windows runners (in a separate commit).

oleg-jukovec commented on net.go:
```diff
@@ -15,22 +15,23 @@
 type listener struct {
 	net.Listener
-	ctx *Ctx
+	sslCtx *Ctx
```
Let's revert this unnecessary change.

DerekBum replied:
Updated.
